### PR TITLE
all.equal dispatches to column methods

### LINKED ISSUE #4541: rbindlist gives wrong results when it is applied to difftime objects with different units.
`#` [`Minimal reproducible example`](https://stackoverflow.com/questions/5963269/how-to-make-a-great-r-reproducible-example)

```
library(data.table)

difftime_diffunits <- list(
  structure(list(
    start_time = "2020-06-02 05:12:14.652", end_time = "2020-06-02 06:01:24.753", 
    diff_time = structure(49.1683500011762, class = "difftime", units = "mins")), 
    row.names = c(NA, -1L), class = c("data.table", "data.frame")), 
  structure(list(
    start_time = "2020-06-02 05:12:14.656", end_time = "2020-06-02 06:02:25.045", 
    diff_time = structure(50.1731500029564, class = "difftime", units = "mins")), 
    row.names = c(NA, -1L), class = c("data.table", "data.frame")), 
  structure(list(
    start_time = "2020-06-02 05:12:14.973", end_time = "2020-06-02 06:32:28.081", 
    diff_time = structure(1.33697444445557, class = "difftime", units = "hours")), 
    row.names = c(NA, -1L), class = c("data.table", "data.frame")), 
  structure(list(
    start_time = "2020-06-02 05:12:14.973", end_time = "2020-06-02 06:24:51.367", 
    diff_time = structure(1.21010944445928, class = "difftime", units = "hours")), 
    row.names = c(NA, -1L), class = c("data.table", "data.frame")))

# [[1]]
# start_time                end_time     diff_time
# 1: 2020-06-02 05:12:14.652 2020-06-02 06:01:24.753 49.16835 mins
# 
# [[2]]
# start_time                end_time     diff_time
# 1: 2020-06-02 05:12:14.656 2020-06-02 06:02:25.045 50.17315 mins
# 
# [[3]]
# start_time                end_time      diff_time
# 1: 2020-06-02 05:12:14.973 2020-06-02 06:32:28.081 1.336974 hours
# 
# [[4]]
# start_time                end_time      diff_time
# 1: 2020-06-02 05:12:14.973 2020-06-02 06:24:51.367 1.210109 hours

rbindlist(difftime_diffunits)

# start_time                end_time      diff_time
# 1: 2020-06-02 05:12:14.652 2020-06-02 06:01:24.753 49.168350 mins
# 2: 2020-06-02 05:12:14.656 2020-06-02 06:02:25.045 50.173150 mins
# 3: 2020-06-02 05:12:14.973 2020-06-02 06:32:28.081  1.336974 mins
# 4: 2020-06-02 05:12:14.973 2020-06-02 06:24:51.367  1.210109 mins
```
**Note** that the units of diff_time of the 3rd and 4th rows should be "**hours**" instead of "**mins**".

`#` `Output of sessionInfo()`
```
R version 4.0.0 (2020-04-24)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows Server x64 (build 17763)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.12.9

loaded via a namespace (and not attached):
[1] compiler_4.0.0 tools_4.0.0    tinytex_0.23   xfun_0.14     
```

### LINKED ISSUE #4543: all.equal.data.table uses relative difference on POSIXct column (should be absolute)
```
DT1 = data.table(t = .POSIXct(1590973200, tz='UTC'))
DT2 = data.table(t = .POSIXct(1590974413, tz='UTC'))
all.equal(DT1, DT2)
# [1] "Column 't': Mean relative difference: 7.624264e-07"
```

vs `base`:

```
DT1 = data.frame(t = .POSIXct(1590973200, tz='UTC'))
DT2 = data.frame(t = .POSIXct(1590974413, tz='UTC'))
all.equal(DT1, DT2)
# [1] "Component “t”: Mean absolute difference: 1213"
```

For `POSIXct` column, relative difference is meaningless since it's somewhat arbitrarily relative to the 0 time of UTC epoch.

### PR DESCRIPTION
**Pull request by MichaelChirico**
Closes #4543 

A bit clunky, open to suggestions for other ways to solve #4543 and also keep test 1710.5.

The improvement here has flushed out some really long-time broken tests in our suite of two flavors:

 - Comparisons on integer64 columns were _really_ broken. Actually we were comparing the underlying double representations, which are typically size 1e-300 floats, which of course "look the same" from `all.equal.numeric()` PoV. Really obvious mismatches were not caught.
 - Comparisons on POSIXct columns that differ in milliseconds. We might prefer to ignore such tiny differences, but I found in our cases it is a meaningful difference useful for testing when `fwrite()` retains sub-second representations vs. not.

Overall I think this is a really nice catch by this PR & glad the suite is improved.

### CODE REVIEW

File: R/setops.R [hunk 1]
```diff
@@ -254,7 +254,7 @@ all.equal.data.table = function(target, current, trim.levels=TRUE, check.attribu
       # trim.levels moved here
       x = target[[i]]
       y = current[[i]]
-      if (xor(is.factor(x),is.factor(y)))
+      if ((is.factor(x) || is.factor(y)) && !(is.factor(x) && is.factor(y)))
```
Comment (MichaelChirico) [inline R/setops.R:257]: `base::xor` uses vector versions: `(x | y) & !(x & y)`

**Comment by jangorecki** [comment]
any idea if this can be a breaking change for any proper use of all.equal?

**Comment by MichaelChirico** [comment]
I tried to guard against that but I can't say for 100% certain... perhaps best to slip this release

**Comment by jangorecki** [comment]
how is this related to rbindlist? I think you wanted to link #4543 @MichaelChirico

**Comment by MichaelChirico** [comment]
Yes thanks @jangorecki

**Comment by MichaelChirico** [comment]
I think this can be included in 1.16.0 👍

File: R/setops.R [hunk 2]
```diff
@@ -282,8 +282,12 @@ all.equal.data.table = function(target, current, trim.levels=TRUE, check.attribu
           # dealt with according to trim.levels
         }
       } else {
-        cols.r = all.equal(unclass(x), unclass(y), tolerance=tolerance, ..., check.attributes=check.attributes)
-        # classes were explicitly checked earlier above, so ignore classes here.
+        # for test 1710.5 and #4543, we want to (1) make sure we dispatch to
+        #   any existing all.equal methods for x while also (2) treating class(x)/class(y)
+        #   as attributes as regards check.attributes argument
+        cols.r = all.equal(x, y, tolerance=tolerance, ..., check.attributes=check.attributes)
+        if (!isTRUE(cols.r) && !check.attributes && isTRUE(all.equal(unclass(x), unclass(y), tolerance=tolerance, ..., check.attributes=check.attributes)))
```
Comment (ben-schwen) [inline R/setops.R:289]:
```suggestion
        if (!isTRUE(cols.r) && !check.attributes && isTRUE(all.equal(unclass(x), unclass(y), tolerance=tolerance, ..., check.attributes=FALSE)))
``` would be clearer IMO

Comment (MichaelChirico) [reply R/setops.R:289]: thanks, I had a similar thought...

File: R/setops.R [hunk 3]
```diff
@@ -282,8 +282,13 @@ all.equal.data.table = function(target, current, trim.levels=TRUE, check.attribu
...
         }
       } else {
-        cols.r = all.equal(unclass(x), unclass(y), tolerance=tolerance, ..., check.attributes=check.attributes)
-        # classes were explicitly checked earlier above, so ignore classes here.
+        # for test 1710.5 and #4543, we want to (1) make sure we dispatch to
+        #   any existing all.equal methods for x while also (2) treating class(x)/class(y)
+        #   as attributes as regards check.attributes argument
+        cols.r = all.equal(x, y, tolerance=tolerance, ..., check.attributes=check.attributes)
+        if (!isTRUE(cols.r) && !check.attributes && isTRUE(all.equal(unclass(x), unclass(y), tolerance=tolerance, ..., check.attributes=FALSE)))
+``` would be clearer IMO
```
Comment (MichaelChirico) [inline R/setops.R:290]:
```suggestion
```

**Comment by MichaelChirico** [comment]
It might be preferable to split off the test fixes that are not specifically related to this PR (but were identified as bugs by the code changes here); let me know.

**Comment by ben-schwen** [comment]
> It might be preferable to split off the test fixes that are not specifically related to this PR (but were identified as bugs by the code changes here); let me know.

Nice! Change looks good, but I think the broken tests warrant a NEWS item now (also since there is a slight change of behavior).

**Comment by MichaelChirico** [comment]
Oh, good catch. It needed a NEWS item anyway.

**Review by ben-schwen — APPROVED** [review]

**Comment by jangorecki** [comment]
> open to suggestions for other ways to solve #4543 and also keep test 1710.5.
```r
test(1710.5, isTRUE(all.equal(x,y,check.attributes=FALSE)))
```

@MichaelChirico I think keeping 1710.5 was a mistake, at least if we aim for base R behavior
```r
x = data.frame(a=0L)
y = data.frame(a=0)
setattr(y[[1]],"class",c("hello","world"))
all.equal(x,y,check.attributes=FALSE)
#[1] "Component “a”: target is numeric, current is hello"
x = as.data.table(x)
y = as.data.table(y)
all.equal(x,y,check.attributes=FALSE)
#[1] TRUE
```
I am now working towards #7434 and would like ideally to change this test.

**Pull request merged**